pkg/replay.go

```diff
@@ -20,97 +20,118 @@ package ripley
+type Options struct {
+	Pace                string
+	Silent              bool
+	SilentHttpError     bool
+	DryRun              bool
+	Timeout             int
+	TimeoutConnection   int
+	Strict              bool
+	Memprofile          string
+	NumWorkers          int
+	PrintStat           bool
+	MetricsServerEnable bool
+	MetricsServerAddr   string
+	NlongestPrint       bool
+	NlongestResults     int
+}
+
+// Ensures we have handled all HTTP request results before exiting
+var waitGroupResults sync.WaitGroup
+
+func Replay(opts *Options) int {
 	// Default exit code
 	var exitCode int = 0
-	// Ensures we have handled all HTTP request results before exiting
-	var waitGroup sync.WaitGroup
 
 	// Send requests for the HTTP client workers to pick up on this channel
-	requests := make(chan *request)
+	requests := make(chan *Request, opts.NumWorkers)
 	defer close(requests)
 
 	// HTTP client workers will send their results on this channel
-	results := make(chan *Result)
+	results := make(chan *Result, opts.NumWorkers)
 	defer close(results)
 
 	// The pacer controls the rate of replay
-	pacer, err := newPacer(phasesStr)
-
+	pacer, err := newPacer(opts.Pace)
 	if err != nil {
 		panic(err)
 	}
 
 	// Read request JSONL input from STDIN
-	scanner := bufio.NewScanner(os.Stdin)
+	reader := bufio.NewReaderSize(os.Stdin, 1024*1024)
+	scanner := bufio.NewScanner(reader)
 
 	// Start HTTP client goroutine pool
-	startClientWorkers(numWorkers, requests, results, dryRun, timeout)
+	startClientWorkers(opts, requests, results)
 	pacer.start()
 
 	for scanner.Scan() {
-		req, err := unmarshalRequest(scanner.Bytes())
+		if pacer.done {
+			break
+		}
+		waitGroupResults.Add(1)
+
+		b := scanner.Bytes()
+		req, err := unmarshalRequest(&b)
 		if err != nil {
 			exitCode = 126
-			result, _ := json.Marshal(Result{
-				StatusCode: 0,
+			res := &Result{
+				StatusCode: -2,
 				Latency:    0,
 				Request:    req,
 				ErrorMsg:   fmt.Sprintf("%v", err),
-			})
-			fmt.Println(string(result))
+			}
 
-			if strict {
+			if opts.Strict {
 				panic(err)
 			}
+
+			results <- res
 			continue
 		}
 
-		if pacer.done {
-			break
-		}
+		duration := pacer.waitDuration(req.Timestamp)
+		time.Sleep(duration)
 
-		// The pacer decides how long to wait between requests
-		waitDuration := pacer.waitDuration(req.Timestamp)
-		time.Sleep(waitDuration)
 		requests <- req
-		waitGroup.Add(1)
-
-		// Goroutine to handle the  HTTP client result
-		go func() {
-			defer waitGroup.Done()
-
-			result := <-results
-
-			// If there's a panic elsewhere, this channel can return nil
-			if result == nil {
-				return
-			}
+	}
 
-			jsonResult, err := json.Marshal(result)
+	if err := scanner.Err(); err != nil {
+		panic(err)
+	}
 
-			if err != nil {
-				panic(err)
-			}
+	waitGroupResults.Wait()
 
-			if !silent {
-				fmt.Println(string(jsonResult))
-			}
-		}()
+	select {
+	case <-metricsRequestReceived:
```

Comment:
I was confused as to where this came from and had to search for the global var in metrics.go. I've left a comment about it there.